{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import csv\n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "find genes wthin sites and claculate amino change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "fp=\"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/nodups_cov5_HE_PLUS_STAR100len/Edits_only/VEP_out.csv\"\n",
    "vep = pd.read_csv(filepath_or_buffer=fp, sep=\"\\t\")\n",
    "#fix location column\n",
    "vep[[\"chr\",\"Location\",]]=vep[\"Location\"].str.split(pat=\":\",expand=True,n=-1)\n",
    "vep[[\"garb\",\"Location\",]]=vep[\"Location\"].str.split(pat=\"-\",expand=True,n=-1)\n",
    "vep.drop(['garb'],axis=1)\n",
    "#fix type\n",
    "vep[\"Location\"] = vep[\"Location\"].astype('int')\n",
    "# remove not refseq-curated sites \n",
    "vep =vep[vep['Feature'].str.startswith('NM')]\n",
    "# remove none cds sites\n",
    "vep =vep[vep['CDS_position'] != '-']\n",
    "\n",
    "\n",
    "# match edit sites to genes\n",
    "edit_site_path = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/nodups_cov5_HE_PLUS_STAR100len/Edits_only/all_uniq_by_frequncy_sites.known.tab\"\n",
    "# list for all sites with their change\n",
    "list_sites_with_genes = []\n",
    "with open(edit_site_path,\"r\") as geneFd:\n",
    "    DictReader_obj = csv.DictReader(geneFd, delimiter=\"\\t\")\n",
    "    for site in DictReader_obj:\n",
    "        site['Coverage-q25'] = int(float(site['Coverage-q25']))\n",
    "        site_vep = vep[vep[\"Location\"] == int(site[\"Position\"])]\n",
    "        # row for every gene symbol \n",
    "        for gene_symb in site_vep['SYMBOL'].unique():\n",
    "            symb_site_vep = site_vep[site_vep[\"SYMBOL\"] == gene_symb]\n",
    "            # will save all the site data even after closing the file\n",
    "            site_with_changes = site.copy()\n",
    "            site_with_changes['mut_type'] = []\n",
    "            # for the site's genes acc num\n",
    "            site_with_changes['genes'] = []\n",
    "            # for the site's amino change per gene\n",
    "            site_with_changes['aminos_change'] =[]\n",
    "            site_with_changes['gene_name']=symb_site_vep.loc[symb_site_vep.index[0],'SYMBOL']\n",
    "            # iterate rows - adding results for every isoform to the symbol row\n",
    "            for index, gene in symb_site_vep.iterrows():\n",
    "                site_with_changes['genes'].append(gene['Feature'])\n",
    "                site_with_changes['mut_type'].append(gene['Consequence'])\n",
    "                if '/' in gene['Amino_acids']:\n",
    "                    site_with_changes['aminos_change'].append(gene['Amino_acids'].replace('/', gene['Protein_position']))\n",
    "                else:\n",
    "                    site_with_changes['aminos_change'].append(gene['Amino_acids']+ gene['Protein_position']+gene['Amino_acids'])\n",
    "            list_sites_with_genes.append(site_with_changes)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/nodups_cov5_HE_PLUS_STAR100len/comfort_ver_final_table_amino_change_by_site_byFreq.tab.csv\"\n",
    "with open(out_file, \"w\") as OFD:\n",
    "    writer = csv.DictWriter(OFD, delimiter=\"\\t\",fieldnames=list(list_sites_with_genes[0].keys()))\n",
    "    writer.writeheader()\n",
    "    writer.writerows(list_sites_with_genes) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for site in list_sites_with_genes:\n",
    "    site['mut_type']=\";\".join(site['mut_type'])\n",
    "    site['aminos_change']=\";\".join(site['aminos_change'])\n",
    "    site['genes']=\";\".join(site['genes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/final_table_amino_change_by_site_byFreq.tab.csv\"\n",
    "with open(out_file, \"w\") as OFD:\n",
    "    writer = csv.DictWriter(OFD, delimiter=\"\\t\",fieldnames=list(list_sites_with_genes[0].keys()))\n",
    "    writer.writeheader()\n",
    "    writer.writerows(list_sites_with_genes) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "out_file_csv = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/final_table_amino_change_by_site_byFreq.csv\"\n",
    "with open(out_file_csv, \"w\") as OFD:\n",
    "    writer = csv.DictWriter(OFD, delimiter=\",\",fieldnames=list(list_sites_with_genes[0].keys()))\n",
    "    writer.writeheader()\n",
    "    writer.writerows(list_sites_with_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# out_file = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/final_table_one_gene_amino_change_by_site_byFreq.tab.csv\"\n",
    "# one_gene_list = []\n",
    "# for site in list_sites_with_genes:\n",
    "#     site_copy = site.copy()\n",
    "#     gene = site_copy['genes'][0]\n",
    "#     mut_t = site_copy['mut_type'][0]\n",
    "#     am_c = site_copy['aminos_change'][0]\n",
    "#     for i in range(len(site['genes'])):\n",
    "#         if site['mut_type'][i] != 'synonymous_variant':\n",
    "#             gene = site_copy['genes'][i]\n",
    "#             mut_t = site_copy['mut_type'][i]\n",
    "#             am_c = site_copy['aminos_change'][i]\n",
    "#             break\n",
    "#     site_copy['genes'] = gene\n",
    "#     site_copy['mut_type'] = mut_t\n",
    "#     site_copy['aminos_change'] = am_c\n",
    "#     one_gene_list.append(site_copy)\n",
    "#     site['mut_type']=\";\".join(site['mut_type'])\n",
    "#     site['aminos_change']=\";\".join(site['aminos_change'])\n",
    "#     site['genes']=\";\".join(site['genes'])\n",
    "# with open(out_file, \"w\") as OFD:\n",
    "#     writer = csv.DictWriter(OFD, delimiter=\"\\t\",fieldnames=list(one_gene_list[0].keys()))\n",
    "#     writer.writeheader()\n",
    "#     writer.writerows(one_gene_list) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "write results to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ALL THE SLASHED CODE BELOW IS NICE CODE BUT ITS NEED GENES' SEQ TABLE WITH UPPERCASE CDS (TABLE BROWSER DIDNT WORK FOR THAT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# type_of_modifications=['AC','AG','AT','CA','CG','CT','GA','GC','GT','TA','TC','TG']\n",
    "# standard_genetic_code = {\n",
    "#     'AUA': 'I', 'AUC': 'I', 'AUU': 'I', 'AUG': 'M',\n",
    "#     'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACU': 'T',\n",
    "#     'AAC': 'N', 'AAU': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "#     'AGC': 'S', 'AGU': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "#     'CUA': 'L', 'CUC': 'L', 'CUG': 'L', 'CUU': 'L',\n",
    "#     'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCU': 'P',\n",
    "#     'CAC': 'H', 'CAU': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "#     'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGU': 'R',\n",
    "#     'GUA': 'V', 'GUC': 'V', 'GUG': 'V', 'GUU': 'V',\n",
    "#     'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCU': 'A',\n",
    "#     'GAC': 'D', 'GAU': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "#     'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGU': 'G',\n",
    "#     'UCA': 'S', 'UCC': 'S', 'UCG': 'S', 'UCU': 'S',\n",
    "#     'UUC': 'F', 'UUU': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "#     'UAC': 'Y', 'UAU': 'Y', 'UAA': None, 'UAG': None,\n",
    "#     'UGC': 'C', 'UGU': 'C', 'UGA': None, 'UGG': 'W'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # read genes seq data\n",
    "# genes_seq_path = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/znf_seq_all_curated.txt\"\n",
    "# all_genes = {}\n",
    "# # create map of genes by refseq gen name(acc num)\n",
    "# with open(genes_seq_path,\"r\") as genesFD:\n",
    "#     for record in SeqIO.parse(genesFD, \"fasta\"):\n",
    "#         temp = record.description.split(\" \")\n",
    "#         # extract name in two steps to ensure success\n",
    "#         name = (temp[0]).split(\"ncbiRefSeqCurated_\")\n",
    "#         assert len(name) > 1, (\"error name pasring :\"+ str(record.description))\n",
    "#         name = name[-1]\n",
    "#         t = temp[1].split(\":\")\n",
    "#         region =t[0].split(\"=\")[1]\n",
    "#         start = t[1].split(\"-\")[0]\n",
    "#         end = t[1].split(\"-\")[1]\n",
    "#         strand = temp[4].split(\"=\")[1]\n",
    "#         seq = str(record.seq)\n",
    "#         all_genes[name] = {\"Name\" : name,\"Region\" : region,  \"Start\" : int(start),\"End\" : int(end),\"Strand\" : strand,\"Seq\" : seq}\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "COUNT OVERLAPPING GENES ON OPPOSITE STRANDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# genes_list = [gene for gene in all_genes.values()]\n",
    "# counter = 0\n",
    "# old1=\"\"\n",
    "# old2=\"\"\n",
    "# for i in range(len(genes_list)):\n",
    "#     for j in range(i, len(genes_list)):\n",
    "#         # for two different genes\n",
    "#         if i != j:\n",
    "#             gene1 = genes_list[i]\n",
    "#             gene2 = genes_list[j]\n",
    "#             # if they are overllaping\n",
    "#             if (gene1[\"Start\"] <= gene2[\"End\"] and gene1[\"End\"] >= gene2[\"End\"]) or (gene1[\"Start\"] <= gene2[\"Start\"] and gene1[\"End\"] >= gene2[\"Start\"]):\n",
    "#                 # if they are on opposite strands\n",
    "#                 if gene1[\"Strand\"] != gene2[\"Strand\"]:\n",
    "#                     #if both genes are new\n",
    "#                     if gene1['Name'] != old1 and gene2['Name'] != old2:\n",
    "#                         old1=gene1['Name']\n",
    "#                         old2=gene2['Name']\n",
    "#                         counter+=1\n",
    "# print(\"overlapping events: \", counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "find duplicated points in REDI input points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !!!!!!!!! very long running time - I found no points\n",
    "# Mf = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/input/Final_file_All_ZNF_A_RefseqCur_knownFormat_sorted.bed\"\n",
    "# with open(Mf, 'r') as FD:\n",
    "#     lines_splited=[line.split(\"\\t\") for line in FD.readlines()]\n",
    "#     for i in range(len(lines_splited)):\n",
    "#         for j in range(i+1,len(lines_splited)):\n",
    "#             if lines_splited[i][0] == lines_splited[j][0] and lines_splited[i][1] == lines_splited[j][1]:\n",
    "#                 print (lines_splited[i], lines_splited[j], \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # match edit sites to genes\n",
    "# edit_site_path = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/curated/nodups_cov5_HE_PLUS_STAR100len/Edits_only/debug_sites.known.tab\"\n",
    "# # list for all sites with their change\n",
    "# list_sites_with_genes = []\n",
    "# with open(edit_site_path,\"r\") as geneFd:\n",
    "#     DictReader_obj = csv.DictReader(geneFd, delimiter=\"\\t\")\n",
    "#     for site in DictReader_obj:\n",
    "#         site['Coverage-q25'] = int(float(site['Coverage-q25']))\n",
    "#         site['contain_nonsyn'] = 0\n",
    "#         # for the site's genes id\n",
    "#         site_genes = []\n",
    "#         # for the site's amono change per gene\n",
    "#         site_aminos_change =[]\n",
    "#         # will save all the site data even after closing the file\n",
    "#         site_with_changes = site.copy()\n",
    "#         #find genes wthin the site\n",
    "#         for name,gene in all_genes.items():\n",
    "#             # if we find a gene\n",
    "#             if site[\"Region\"] == gene[\"Region\"] and gene[\"Start\"] < int(site[\"Position\"]) < gene[\"End\"]:\n",
    "#                 # calc relative position according to strand\n",
    "#                 gene_seq=gene[\"Seq\"]\n",
    "#                 print(gene_seq)\n",
    "#                 # !note! genes seq is already in reverse complement\n",
    "#                 # !note! that reverse complement will convert CCTGTTC to GAACAGG (complement + reverse) - so the different in strands is only calculation of starting point\n",
    "#                 if site[\"Strand\"] == \"1\": # strand is +\n",
    "#                     gene_relative_site = int(site[\"Position\"]) - gene[\"Start\"]\n",
    "#                     # make sure editing and gene on the same strand\n",
    "#                     if gene[\"Strand\"] != \"+\":\n",
    "#                         continue   \n",
    "#                 else:\n",
    "#                     gene_relative_site = gene[\"End\"] - int(site[\"Position\"])\n",
    "#                     # make sure editing and gene on the same strand\n",
    "#                     if gene[\"Strand\"] != \"-\":\n",
    "#                         continue   \n",
    "#                 # make sure site is wthin codin region\n",
    "#                 if gene_seq[gene_relative_site].islower():\n",
    "#                     continue\n",
    "                                 \n",
    "#                 assert gene_seq[gene_relative_site] == \"A\", \"no A,\" + str(site) + \"gene :\"+ name\n",
    "                \n",
    "#                 ## for case of other replacement\n",
    "#                 ## should be removed in the \"collect recurrent...ipynb\" file\n",
    "#                 # if gene_seq[gene_relative_site] != \"A\":\n",
    "#                 #     assert \"AG\" not in site[\"AllSubs\"], \"AG in subs but not in index \" + gene[\"Seq\"] + \"strand: \" + gene[\"Strand\"]+site[\"Strand\"]+\" got \"+gene_seq[gene_relative_site] + \"area :\" + gene_seq[gene_relative_site-3:gene_relative_site+4] +\" site: \" +str(site)\n",
    "#                 #     continue\n",
    "                \n",
    "#                 #update sitte genes       \n",
    "#                 site_genes.append(name)\n",
    "#                 site_with_changes[\"Genes\"] = site_genes\n",
    "#                 #->find amino replacemant<-\n",
    "#                 # will point where we are in the list \n",
    "#                 idx = 0\n",
    "#                 # will cout how many aminos so far\n",
    "#                 amino_counter =1 \n",
    "#                 border = gene_relative_site+1\n",
    "                \n",
    "#                 #find amino change seconded versions\n",
    "#                 #find where our condon ended\n",
    "#                 temp_seq = gene_seq[:gene_relative_site+1]\n",
    "#                 upps = len(re.findall(r'[A-Z]',temp_seq))\n",
    "#                 mdl=upps%3\n",
    "#                 if mdl==0:\n",
    "#                     adding=0\n",
    "#                 else:\n",
    "#                     adding=3-mdl\n",
    "#                 relvant_seq=gene_seq[:gene_relative_site+adding+1]\n",
    "#                 upps=len(re.findall(r'[A-Z]',relvant_seq))\n",
    "#                 assert upps%3==0, \"%3!=0\"\n",
    "#                 amino_counter=int(upps/3)\n",
    "#                 # take last 3 chars\n",
    "#                 seq_relvant_seq=MutableSeq(relvant_seq)\n",
    "#                 codon=seq_relvant_seq[-3:]\n",
    "#                 assert str(codon).isupper(), \"CODON not upper\"+codon\n",
    "#                 orig_amino=str(codon.translate())\n",
    "#                 seq_relvant_seq[gene_relative_site]=\"G\"\n",
    "#                 new_amino=str(seq_relvant_seq[-3:].translate())\n",
    "#                 if orig_amino != new_amino:\n",
    "#                             site_with_changes['contain_nonsyn'] = 1\n",
    "#                 site_aminos_change.append(orig_amino+str(amino_counter)+new_amino)\n",
    "#                 site_with_changes[\"Amino_change\"] = site_aminos_change\n",
    "                \n",
    "                \n",
    "#                 # while(idx < border):       \n",
    "#                 #         # if the site is in the current 3 pos\n",
    "#                 #     if idx+3 > gene_relative_site:\n",
    "#                 #         # move forward untill its upper \n",
    "#                 #         while gene_seq[idx].islower():\n",
    "#                 #             idx+=1\n",
    "#                 #         # check we didnt pass the site\n",
    "#                 #         assert idx <= gene_relative_site, \"idx > gene_relative_site\"\n",
    "#                 #         codon = gene_seq[idx:idx+3]\n",
    "#                 #         # check its three coding nucleotide\n",
    "#                 #         assert codon[idx:idx+3].isupper, \"codon contain lower-cases\" + str(codon)\n",
    "#                 #         # check again its A - otherwise its error in this indexses part\n",
    "#                 #         assert codon[gene_relative_site - idx] == \"A\", \"error codon no A in place \"+str(codon)+ \" - \" +gene_seq[gene_relative_site-2:gene_relative_site+1]+\" > site \" + str(gene_relative_site) + \" idx \"+  str(idx)\n",
    "#                 #         codon = Seq(codon)\n",
    "#                 #         orig_amino = str(codon.translate())\n",
    "#                 #         new_seq = MutableSeq(codon)\n",
    "#                 #         new_seq[gene_relative_site - idx] = \"G\"\n",
    "                        \n",
    "#                 #         new_amino = str(Seq(new_seq).translate())\n",
    "#                 #         if orig_amino != new_amino:\n",
    "#                 #             site_with_changes['contain_nonsyn'] = 1\n",
    "#                 #         site_aminos_change.append(orig_amino+str(amino_counter)+new_amino)\n",
    "\n",
    "#                 #         site_with_changes[\"Amino_change\"] = site_aminos_change\n",
    "#                 #         #finish runing on the seq\n",
    "#                 #         break\n",
    "#                 #     elif gene_seq[idx:idx+3].isupper():\n",
    "#                 #         amino_counter += 1\n",
    "#                 #         idx +=3\n",
    "#                 #     elif  gene_seq[idx].islower():    #if its lower letter move forward\n",
    "#                 #         idx+=1\n",
    "#                 #     else:\n",
    "#                 #         raise(\"somthig wrong with orf\") \n",
    "\n",
    "#         # add site to list\n",
    "#         list_sites_with_genes.append(site_with_changes)\n",
    "\n",
    "# # keep sites wthin CDS only(remove only 4 sites)\n",
    "# list_sites_with_genes = [site for site in list_sites_with_genes if 'Genes' in site.keys()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "add genes names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_dict_file = \"/private8/Projects/itamar/ZNF/Sra_GSE73211_35_samples/REDI/refseq_curated_ZNF_genes_name_dcit.txt\"\n",
    "\n",
    "# with open(gene_dict_file,\"r\") as genes_names_Fd:\n",
    "#     DictReader_obj = csv.DictReader(genes_names_Fd, delimiter=\"\\t\")\n",
    "#     genes_names_dict = {line[\"#name\"]: line[\"name2\"]  for line in DictReader_obj}\n",
    "\n",
    "# for site in list_sites_with_genes:\n",
    "#     site[\"gene_name\"] = set([])\n",
    "#     for gene_id in site[\"Genes\"]:\n",
    "#         site[\"gene_name\"].add(genes_names_dict[gene_id])\n",
    "# for site in list_sites_with_genes:\n",
    "#     site[\"gene_name\"]=list(site[\"gene_name\"])\n",
    "#     if len(site[\"gene_name\"]) > 1:\n",
    "#         if \"-\" in site[\"gene_name\"][0]:\n",
    "#             site[\"gene_name\"] = site[\"gene_name\"][1]\n",
    "#         else:\n",
    "#             site[\"gene_name\"] = site[\"gene_name\"][0]\n",
    "# # make data readable\n",
    "# for site in list_sites_with_genes:\n",
    "#     site[\"Genes\"]= \";\".join(site[\"Genes\"])\n",
    "#     site[\"gene_name\"]= \"\".join(site[\"gene_name\"])\n",
    "#     site['Amino_change'] = \";\".join(site['Amino_change'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('lamootHW')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3566c55396aa67d6d450caee2c72072e0a80bcedc801748038b289e6d3e79506"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
